src/landlab/utils/channel_network_grid_tools.py

```diff
@@ -0,0 +1,368 @@
+import itertools
+
+import numpy as np
+import pandas as pd
+
+from landlab.components.flow_director.flow_director_steepest import FlowDirectorSteepest
+
+"""
+A collection of tools for mapping values (e.g., flow, shear stress) between a
+network model grid and raster model grid representation of a channel network.
+"""
+
+
+def _flatten_lol(lol):
+    """Flatten a list of lists.
+
+    Parameters
+    ----------
+    lol : list of list
+        A list where each element is itself a list.
+
+    Returns
+    -------
+    list
+        A single list containing all elements from the sublists.
+    """
+    return list(itertools.chain.from_iterable(lol))
+
+
+def get_link_nodes(nmgrid):
+    """Get the downstream (head) and upstream (tail) nodes at a link from
+    flow director. The network model grid nodes_at_link attribute may not be
+    ordered according to flow direction. Output from this function should be
+    used for all channel_network_grid_tools functions that require a link_nodes input
+
+    Parameters
+    ----------
+    nmgrid : network model grid
+
+    Returns
+    -------
+    link_nodes : np.array
+        for a nmgrid of n nodes, returns a nx2 np array, the ith row of the
+        array is the [downstream node id, upstream node id] of the ith link
+    """
+
+    fd = FlowDirectorSteepest(nmgrid, "topographic__elevation")
+    fd.run_one_step()
+    upstream_node_id = []
+    downstream_node_id = []
+    for i in range(nmgrid.number_of_links):
+        upstream_node_id.append(fd.upstream_node_at_link()[i])
+        downstream_node_id.append(fd.downstream_node_at_link()[i])
+    # create np array and transpose, each row is [downstream node id, upstream node id]
+    link_nodes = np.array([downstream_node_id, upstream_node_id]).T
+    return link_nodes
```

Comment:
I think this can be simplified to,
```python
    return np.column_stack(
        (fd.downstream_node_at_link(), fd.upstream_node_at_link())
    ).astype(int, copy=False)
```

Reply:
changed, thanks